hyperspy/io_plugins/emd.py

```diff
@@ -0,0 +1,392 @@
+    comments: dictionary
+        Dictionary which contains additional commentary metadata.
+
+    """
+
+    _log = logging.getLogger(__name__)
+
+    def __init__(self, signals=None, user=None, microscope=None, sample=None, comments=None):
+        self._log.debug('Calling __init__')
+        # Create dictionaries if not present:
+        if signals is None:
+            signals = {}
+        if user is None:
+            user = {}
+        if microscope is None:
+            microscope = {}
+        if sample is None:
+            sample = {}
+        if comments is None:
+            comments = {}
+        # Make sure some default keys are present in user:
+        for key in ['name', 'institution', 'department', 'email']:
+            if key not in user:
+                user[key] = ''
+        self.user = user
+        # Make sure some default keys are present in microscope:
+        for key in ['name', 'voltage']:
+            if key not in microscope:
+                microscope[key] = ''
+        self.microscope = microscope
+        # Make sure some default keys are present in sample:
+        for key in ['material', 'preparation']:
+            if key not in sample:
+                sample[key] = ''
+        self.sample = sample
+        # Add comments:
+        self.comments = comments
+        # Make sure the signals are added properly to signals:
+        self.signals = {}
+        for name, signal in signals.iteritems():
+            self.add_signal(signal, name)
+
+    def __getitem__(self, key):
+        # This is for accessing the raw data easily. For the signals use emd.signals[key]!
+        return self.signals[key].data
+
+    def _write_signal_to_group(self, signal_group, signal):
+        self._log.debug('Calling _write_signal_to_group')
+        # Save data:
+        dataset = signal_group.create_group(signal.metadata.General.title)
+        maxshape = tuple(None for _ in signal.data.shape)
+        dataset.create_dataset('data', data=signal.data, chunks=True, maxshape=maxshape)
+        # Iterate over all dimensions:
+        for i in range(len(signal.data.shape)):
+            key = 'dim{}'.format(i+1)
+            offset = signal.axes_manager[i].offset
+            scale = signal.axes_manager[i].scale
+            dim = dataset.create_dataset(key, data=[offset, offset+scale])
+            name = signal.axes_manager[i].name
+            from traits.trait_base import _Undefined
+            if type(name) is _Undefined:
+                name = ''
+            dim.attrs['name'] = name
+            units = signal.axes_manager[i].units
+            if type(units) is _Undefined:
+                units = ''
+            else:
+                units = '[{}]'.format('_'.join(list(units)))
+            dim.attrs['units'] = units
+        # Write metadata:
+        dataset.attrs['emd_group_type'] = 1
+        for key, value in signal.metadata.Signal:
+            dataset.attrs[key] = value

```

Comment:
I'll use a `try/except` block for now and let `h5py` handle the rest. My intention was to save everything that is applicable for all signals in the dictionaries on the top level (`user`, `microscope`, etc.) and things which belong to the specific signal (which would be saved in `signal.metadata.Signal`, right?) inside the `dataset` attributes.
I didn't think about the fact that users could store more complex data in `signals.metadata.Signal`, but losing these would not be dramatic, my intention was to just save as much information as possible in the EMD.
If the demand is there, the algorithm could search through the metadata and save everything which is more complex in another dataset (besides `data` and the several `dim` ones which are always present, see the second graph at http://emdatasets.lbl.gov/spec/), but for now I would leave it as it is...